[rustdoc] Don't document stripped items in JSON renderer.

aDotInTheVoid opened the pull request:
Fixes #80664, see [my comment there](https://github.com/rust-lang/rust/issues/80664#issuecomment-797557948) for why

Note that we already do something similar in `convert_item`:

https://github.com/rust-lang/rust/blob/bb4cdf8ec034dca5c056ec9295f38062e5b7e871/src/librustdoc/json/conversions.rs#L28-L31

@rustbot modify labels: +T-rustdoc +A-rustdoc-json

r? @jyn514 
cc @CraftSpider

aDotInTheVoid commented:
This line

https://github.com/rust-lang/rust/blob/b97fd3e5a1545ab02e18c52e7f3d2e78a5c960bf/src/librustdoc/json/conversions.rs#L182

Should be changed to `unreachable` I think, as we should never call `from_clean_item_kind` on a `StrippedItem`
(or make `from_clean_item_kind` return an `Option`)

jyn514 commented on src/librustdoc/formats/renderer.rs:
```diff
@@ -13,6 +13,10 @@ crate trait FormatRenderer<'tcx>: Sized {
 
+    /// Whether to call `item` on `clean::StrippedItem`
+    /// See #80664
+    const RUN_ON_STRIPED: bool;
```
```suggestion
    const RUN_ON_STRIPPED: bool;
```

jyn514 commented on src/librustdoc/formats/renderer.rs:
```diff
@@ -84,13 +88,22 @@ crate fn run_format<'tcx, T: FormatRenderer<'tcx>>(
-            let module = match *item.kind {
-                clean::StrippedItem(box clean::ModuleItem(m)) | clean::ModuleItem(m) => m,
+
+            if let Some(module) = match *item.kind {
+                clean::ModuleItem(m) => Some(m),
+                clean::StrippedItem(box clean::ModuleItem(m)) => {
+                    if T::RUN_ON_STRIPED {
+                        Some(m)
+                    } else {
+                        None
+                    }
+                }
```
nit: I find this pretty hard to read, I would assign the option to a local variable instead of putting a `match` in the if condition.

jyn514 reviewed (approved):
Looks good assuming my comment in https://github.com/rust-lang/rust/issues/80664#issuecomment-803744519 is accurate that JSON doesn't care about inlining.

jyn514 commented:
> This line
> 
> https://github.com/rust-lang/rust/blob/b97fd3e5a1545ab02e18c52e7f3d2e78a5c960bf/src/librustdoc/json/conversions.rs#L182
> 
> Should be changed to `unreachable` I think, as we should never call `from_clean_item_kind` on a `StrippedItem`
> (or make `from_clean_item_kind` return an `Option`)

Sounds good - could you add that here?

jyn514 commented on src/librustdoc/json/mod.rs:
```diff
@@ -170,12 +170,9 @@ impl<'tcx> FormatRenderer<'tcx> for JsonRenderer<'tcx> {
-                assert_eq!(old_item, new_item);
-            }
+            let removed = self.index.borrow_mut().insert(from_def_id(id), new_item);
+
+            assert_matches!(removed, None);
```
```suggestion

            assert!(removed.is_none());
```

aDotInTheVoid replied:
This gives a nicer panic messages when `removed` is `Some`

jyn514 reviewed (commented):
It looks like you're still making changes, but when you're ready for review - what's going on in https://github.com/rust-lang/rust/pull/83055/commits/bdf6786f91a98c702f7ba624b157c1685a6e5db1 ?

aDotInTheVoid commented:
> It looks like you're still making changes, but when you're ready for review - what's going on in [bdf6786](https://github.com/rust-lang/rust/commit/bdf6786f91a98c702f7ba624b157c1685a6e5db1) ?

It turns out that we also don't need to call `item` even for no striped modules. And this fixes the Duplication, at least with the current test suite. I'm trying it with tests for #83057 now, and it looks less promizing.

I'll let you know when this is ready for review

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2170140175) [(plain)](https://github.com/rust-lang/rust/commit/bb2d9537f0496a4bdff81cb4990be2b3cd9321b2/checks/2170140175/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/src/tools/jsondocck/src/main.rs at line 182:
         commands.push(Command { negated, kind: cmd, args, lineno })
 
-    if !errors {
-        Ok(commands)
-    } else {
-    } else {
-        Err(())
-    }
+    if !errors { Ok(commands) } else { Err(()) }
 
 
 /// Performs the actual work of ensuring a command passes. Generally assumes the command
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/src/tools/linkchecker/main.rs" "/checkout/src/tools/error_index_generator/build.rs" "/checkout/src/tools/error_index_generator/main.rs" "/checkout/src/tools/jsondocck/src/config.rs" "/checkout/src/tools/jsondocck/src/error.rs" "/checkout/src/tools/jsondocck/src/cache.rs" "/checkout/src/tools/jsondocck/src/main.rs" "/checkout/src/tools/tidy/src/unit_tests.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
Build completed unsuccessfully in 0:00:14
```

</details>

aDotInTheVoid commented:
This PR is a bit of a mess, but now does 3 things

1. Avoids recursing in `run_format`, controlled by a const in `FormatRenderer`. This fixes #80664. Also add a test for that
2. Adds tests for where the index still has collisions, even with change 1
3. Improvements to `jsondocck` to make the first 2 changes possible

The git history is a mess, and I'd be willing to clean it up and/or split this into 3 pr's if you want

@rustbot modify labels: -S-waiting-on-author +S-waiting-on-review

jyn514 commented:
> It turns out that we also don't need to call item even for non stripped modules. And this fixes the Duplication, at least with the current test suite. I'm trying it with tests for #83057 now, and it looks less promizing.

Can you post what tests fail when you disable this for the HTML backend too? I'd like to keep it in sync with the JSON backend where possible.

jyn514 commented:
> The git history is a mess, and I'd be willing to clean it up and/or split this into 3 pr's if you want

Splitting this into 2 commits - one changing jsondocck, the other fixing the bug - seems fine.

jyn514 commented:
> Adds tests for where the index still has collisions, even with change 1

It wasn't clear to me by reading: which test is this? Where is the collision? Could you add it as a comment?

aDotInTheVoid committed:
Better errors in jsondocck

aDotInTheVoid committed:
Don't call `item` on modules for json renderer

Closes #80664

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-9`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2185803677) [(plain)](https://github.com/rust-lang/rust/commit/eb72f7e71c90684a8ae395b12828e8265d894b07/checks/2185803677/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.................................................................................................... 9400/11713
.................................................................................................... 9500/11713
......................................................i......i...................................... 9600/11713
.................................................................................................... 9700/11713
iiiiiii..iiiiii.i................................................................................... 9800/11713
.................................................................................................... 10000/11713
.................................................................................................... 10100/11713
.................................................................................................... 10200/11713
.................................................................................................... 10300/11713
---
Suite("src/test/assembly") not skipped for "bootstrap::test::Assembly" -- not in ["src/tools/tidy"]
Check compiletest suite=assembly mode=assembly (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 31 tests
iiiiiiiiiiiiiiiiiiiiiiiiiiiii..

 finished in 0.096 seconds
Suite("src/test/incremental") not skipped for "bootstrap::test::Incremental" -- not in ["src/tools/tidy"]
Check compiletest suite=incremental mode=incremental (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
---
Suite("src/test/debuginfo") not skipped for "bootstrap::test::Debuginfo" -- not in ["src/tools/tidy"]
Check compiletest suite=debuginfo mode=debuginfo (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 116 tests
iiiiiiiiii.i.i..i..i..ii....i.i....ii..........iiii.........i.....i...i.......ii.i.ii.....iiii.....i 100/116
test result: ok. 78 passed; 0 failed; 38 ignored; 0 measured; 0 filtered out; finished in 2.27s

 finished in 2.331 seconds
Suite("src/test/ui-fulldeps") not skipped for "bootstrap::test::UiFullDeps" -- not in ["src/tools/tidy"]
---
Some tests failed in compiletest suite=rustdoc-json mode=rustdoc-json host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
F...................
failures:

---- [rustdoc-json] rustdoc-json/reexport/in_root_and_mod_pub.rs stdout ----

error: jsondocck failed!
status: exit code: 101
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/jsondocck" "--doc-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-json/reexport/in_root_and_mod_pub" "--template" "/checkout/src/test/rustdoc-json/reexport/in_root_and_mod_pub.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `0`,
 right: `1`: Didn't get 1 result for `$.index[*][?(@.inner.span=='foo::Bar')].id`: got []', src/tools/jsondocck/src/main.rs:263:13

------------------------------------------




failures:
    [rustdoc-json] rustdoc-json/reexport/in_root_and_mod_pub.rs
test result: FAILED. 19 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 1.62s




command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--rustdoc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "--jsondocck-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/jsondocck" "--src-base" "/checkout/src/test/rustdoc-json" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-json" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "rustdoc-json" "--mode" "rustdoc-json" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-9/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "9.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver engine executionengine fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test --exclude src/tools/tidy
Build completed unsuccessfully in 0:28:44
```

</details>

aDotInTheVoid commented:
Welp, I guess not. Cant reproduce the error (`./x.py test  src/test/rustdoc-json/reexport/in_root_and_mod_pub.rs`) passes

@rustbot modify labels: -S-waiting-on-review +S-waiting-on-author

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-9`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2186285951) [(plain)](https://github.com/rust-lang/rust/commit/9ba92972ed6a1a39afa5993da97e016eea907be2/checks/2186285951/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.................................................................................................... 9400/11713
.................................................................................................... 9500/11713
......................................................i.......i..................................... 9600/11713
.................................................................................................... 9700/11713
iiiiiii..iiiiii.i................................................................................... 9800/11713
.................................................................................................... 10000/11713
.................................................................................................... 10100/11713
.................................................................................................... 10200/11713
.................................................................................................... 10300/11713
---
Suite("src/test/assembly") not skipped for "bootstrap::test::Assembly" -- not in ["src/tools/tidy"]
Check compiletest suite=assembly mode=assembly (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 31 tests
iiiiiiiiiiiiiiiiiiiiiiiiiiiii..

 finished in 0.094 seconds
Suite("src/test/incremental") not skipped for "bootstrap::test::Incremental" -- not in ["src/tools/tidy"]
Check compiletest suite=incremental mode=incremental (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
---
Suite("src/test/debuginfo") not skipped for "bootstrap::test::Debuginfo" -- not in ["src/tools/tidy"]
Check compiletest suite=debuginfo mode=debuginfo (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 116 tests
iiiiiiiiii.i.i..i..i..ii....i.i....ii..........iiii.........i.....i...i.......ii.i.ii.....iiii.....i 100/116
test result: ok. 78 passed; 0 failed; 38 ignored; 0 measured; 0 filtered out; finished in 2.42s

 finished in 2.489 seconds
Suite("src/test/ui-fulldeps") not skipped for "bootstrap::test::UiFullDeps" -- not in ["src/tools/tidy"]
---
Some tests failed in compiletest suite=rustdoc-json mode=rustdoc-json host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
F...................
failures:

---- [rustdoc-json] rustdoc-json/reexport/in_root_and_mod_pub.rs stdout ----

error: jsondocck failed!
status: exit code: 101
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/jsondocck" "--doc-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-json/reexport/in_root_and_mod_pub" "--template" "/checkout/src/test/rustdoc-json/reexport/in_root_and_mod_pub.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `0`,
 right: `1`: Didn't get 1 result for `$.index[*][?(@.inner.span=='foo::Bar')].id`: got []', src/tools/jsondocck/src/main.rs:263:13

------------------------------------------




failures:
    [rustdoc-json] rustdoc-json/reexport/in_root_and_mod_pub.rs
test result: FAILED. 19 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 1.65s




command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--rustdoc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "--jsondocck-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/jsondocck" "--src-base" "/checkout/src/test/rustdoc-json" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-json" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "rustdoc-json" "--mode" "rustdoc-json" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-9/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "9.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver engine executionengine fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test --exclude src/tools/tidy
Build completed unsuccessfully in 0:29:37
```

</details>

aDotInTheVoid commented:
Rebased onto master, can reproduce

aDotInTheVoid committed:
Rename `span` to `source`

Caused by
https://github.com/rust-lang/rust/commit/b0659f9b1bfb92626c40dabceb3268f88bb26224#diff-ede26372490522288745c5b3df2b6b2a1cc913dcd09b29af3a49935afe00c7e6L464-R464

aDotInTheVoid commented:
@rustbot modify labels: -S-waiting-on-author +S-waiting-on-review

jyn514 commented:
@bors r+

bors commented:
:pushpin: Commit d9e2d8d6652fef745a5658178a6f85fc809f6453 has been approved by `jyn514`

<!-- @bors r=jyn514 d9e2d8d6652fef745a5658178a6f85fc809f6453 -->
<!-- homu: {"type":"Approved","sha":"d9e2d8d6652fef745a5658178a6f85fc809f6453","approver":"jyn514"} -->
